ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+
+
+#### JSON Converter
+
+The default JSON converter will respect [the interface](https://ruby-doc.org/stdlib-3.1.2/libdoc/json/rdoc/JSON.html#module-JSON-label-Custom+JSON+Additions)
+provided by the Ruby's JSON module to serialize/deserialize an object.
+
+
+### Codecs
+
+And a payload codec is expected to implement this interface:
+
+```ruby
+interface _PayloadCodec
+  # Takes an array of Payload objects and returns an array of encoded Payload object
+  # The resulting Array is expected to hold at least one Payload
+  def encode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+
+  # Takes an array of Payloads object and returns an array of decoded Payload object
+  # The resulting Array is expected to hold at least one Payload
+  def decode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+end
+```
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  # errors specific to Client
+  class ClientError < Error
+
+  # errors specific to Worker
+  class WorkerError < InternalError
+
+  # superclass for connection, network and response errors
+  class RPCError < Error
+  class UnexpectedError < RPCError
+
+  # superclass for all API failure responses
+  class APIError < ServerError
+
+  # type hinting errors
+  class TypeError < Error
+
+  # superclass for workflow errors
+  class WorkflowError < Error
+```
+
+Notes:
+
+- This list is not exhaustive, more errors will be added as parts of the SDK are getting built
+- Some concrete error class placement might be confusing and will be clarified later in the process
+
+
+### Converting API Failures
+
+Similar to other SDKs we will decode `Failure` proto messages to native error classes. All `cause`s
+will be decoded recursively and the result will form a hierarchy identical to the one in the proto
+response. These error classes will all be subclasses of an `APIError` and will have names matching
+the original protobuf messages (while replacing `Failure` with `Error` for consistency within Ruby):
+
+```ruby
+module Temporal
+  class ApplicationError < APIError
+  class TimeoutError < APIError
+  class CanceledError < APIError
+  class TerminatedError < APIError
+  class ServerError < APIError
+  class ResetWorkflowError < APIError
+  class ActivityError < APIError
+  class ChildWorkflowExecutionError < APIError
+end
+```
+
+At this point we are not going to be converting user errors back to their original classes. These
+will be decoded into `ApplicationError`s while keeping the original class name, message, stack trace
+and other details. Here's an example of handling an activity error:
+
+```ruby
+class ProcessPaymentActivity < Temporal::Activity
+  class InsufficientFunds < Temporal::NonRetryableError; end
```

Comment:
I didn't see `Temporal::NonRetryableError` in the hierarchy before. Can you clarify its superclass?

Reply:
I would keep the other SDK's convention of `ApplicationError.nonRetryable`, it maps directly to the protocol and is easier to spot when inspecting a workflow's history in the UI.

Reply:
@bergundy are you suggesting nesting it further (e.g. `Temporal::ApplicationError::NonRetryable`)? Or something else?

Reply:
No, I'm suggesting that ApplicationError would have a retryable flag.

Reply:
So in order to raise a non-retryable error you'd need to either:

```ruby
class MyError < Temporal::ApplicationError; end
...
raise MyError.new('error message', non_retryable: true)
```

or:

```ruby
class MyError < Temporal::ApplicationError
  def initialize(message)
    super(message, non_retryable: true)
  end
end
...
raise MyError.new('error message')
```

am I interpreting this correctly? As in — there's no simple way of defining a non-retryable error type without additional code?

I still think a superclass approach is better. Maybe we can make it more explicit `Temporal::NonRetryableApplicationError`?

Reply:
This is fine IMHO:

```ruby
class MyError < Temporal::ApplicationError
  def initialize(message)
    super(message, non_retryable: true)
  end
end
```

We also provide factory methods to instantiate retryable and non-retryable application failures in other SDKs.